src/ellswift.rs

```diff
@@ -0,0 +1,704 @@
+    let mut x32_array = [0u8; 32];
+    let mut ell_a64_array = [0u8; 64];
+    let mut ell_b64_array = [0u8; 64];
+
+    // Copy the data into Rust slices
+    ptr::copy_nonoverlapping(x32, x32_array.as_mut_c_ptr(), 32);
+    ptr::copy_nonoverlapping(ell_a64, ell_a64_array.as_mut_c_ptr(), 64);
+    ptr::copy_nonoverlapping(ell_b64, ell_b64_array.as_mut_c_ptr(), 64);
+    // Call the hash function that was passed in through the `data` pointer
+    let secret = callback(x32_array, ell_a64_array, ell_b64_array);
+    // Copy the output from a [ElligatorSwiftSharedSecret] into the output pointer
+    ptr::copy_nonoverlapping(secret.0.as_ptr(), output, secret.0.len());
+
+    secret.0.len() as c_int
+}
+
+/// `ElligatorSwift` is an encoding of a uniformly chosen point on the curve
+/// as a 64-byte array that is indistinguishable from a uniformly random array.
+/// This object holds two field elements u and t, which are the inputs to
+/// the `ElligatorSwift` encoding function.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwift(ffi::ElligatorSwift);
+
+impl ElligatorSwift {
+    /// Create a new `ElligatorSwift` object from a 64-byte array.
+    pub fn new(secret_key: SecretKey, rand: [u8; 32]) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            let ret = ffi::secp256k1_ellswift_create(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                secret_key.as_c_ptr(),
+                rand.as_ptr(),
+            );
+            debug_assert_eq!(ret, 1);
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+
+    /// Creates an `ElligatorSwift` object from a 64-byte array.
+    pub fn from_array(ellswift: [u8; 64]) -> ElligatorSwift {
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ellswift))
+    }
+
+    /// Returns the 64-byte array representation of this `ElligatorSwift` object.
+    pub fn to_array(&self) -> [u8; 64] { self.0.to_array() }
+
+    /// Creates the Elligator Swift encoding from a secret key, using some aux_rand if defined.
+    /// This method is preferred instead of just decoding, because the private key offers extra
+    /// entropy.
+    /// # Example
+    /// ```
+    /// # #[cfg(feature = "alloc")] {
+    ///     use secp256k1::{ellswift::ElligatorSwift, PublicKey, Secp256k1, SecretKey};
+    ///     let secp = Secp256k1::new();
+    ///     let sk = SecretKey::from_slice(&[1; 32]).unwrap();
+    ///     let es = ElligatorSwift::from_seckey(&secp, sk, None);
+    /// # }
+    /// ```
+    pub fn from_seckey<C: Verification>(
+        secp: &Secp256k1<C>,
+        sk: SecretKey,
+        aux_rand: Option<[u8; 32]>,
+    ) -> ElligatorSwift {
+        let mut es_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        let aux_rand_ptr =
+            aux_rand.as_ref().map(|rand| rand.as_c_ptr()).unwrap_or(core::ptr::null());
```

Comment:
In 9b9090b2d777d2dade6f06579fcd0e90a88d8325:

Could you actually just implement `CPtr` for `Option<T>` where `T: CPtr`? Then you could avoid doing this complicated unwrap thing each time.